### PR TITLE
WIP: New detector for anomaly #607

### PR DESCRIPTION
**Pull request by VargBurz**

### CODE REVIEW

File: analytics/analytics/detectors/anomaly_detector.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+class AnomalyDetector(Detector):
+
+    def __init__(self):
+        self.amount = 10
+        pass
+
+    def train(self, dataframe: pd.DataFrame, condfidence: float, cache: Optional[ModelCache]) -> ModelCache:
+        return {
+            'cache': {
```
Comment (jonyrock) [inline analytics/analytics/detectors/anomaly_detector.py]: please don't return `dict` here. Make refactoring for this if it is necessary

File: analytics/analytics/detectors/anomaly_detector.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                '': ,
+                '': ,
+            }
+        }
+
+    def smooth_data(self, dataframe: pd.DataFrame) -> List[List[int, float]]:
```
Comment (jonyrock) [inline analytics/analytics/detectors/anomaly_detector.py]: you should have one method for data preparation and all general algorithms better have in another place.

File: analytics/analytics/detectors/anomaly_detector.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        #smooth data using exponential smoothing/moving average/weighted_average
+        pass
+    
+    def get_confidence_window(self, smooth_data: pd.Series, condfidence: float) -> List[pd.Series, pd.Series]:
+        #build confidence interval above and below smoothed data
+        pass
+    
+    def find_anomaly(self, data, conf):
```
Comment (jonyrock) [inline analytics/analytics/detectors/anomaly_detector.py]: I think you should have only one method `detect` without `find_anomaly`

File: analytics/analytics/detectors/anomaly_detector.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+from typing import Optional, Union
+from analytic_types import AnalyticUnitId
+from detectors import Detector
+
+
+
+class AnomalyDetector(Detector):
+
+    def __init__(self):
+        self.amount = 10
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: Let's name it more clear

File: analytics/analytics/detectors/anomaly_detector.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+logger = logging.getLogger('ANOMALY_DETECTOR')
+
+
+class AnomalyDetector(Detector):
+
+    def __init__(self, *args, **kwargs):
+        self.bucket = DataBucket()
+
+    def train(self, dataframe: pd.DataFrame, payload: Union[list, dict], cache: Optional[ModelCache]) -> ModelCache:
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: Are you sure `payload` can be `list`?

File: analytics/analytics/detectors/anomaly_detector.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            'cache': {
+                'confidence': payload['confidence'],
+                'alpha': payload['alpha']
+            }
+        }
+
+    def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> dict:
+        last_values = None
+        if cache is not None:
+            last_values = cache['last_values']
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: Is `ModelCache` a `dict`? It should be `ModelState`

File: analytics/analytics/detectors/anomaly_detector.py [hunk 7]
```diff
@@ -31,6 +31,14 @@ def get_confidence_window(self, smooth_data: pd.Series, condfidence: float) -> L
     
     def find_anomaly(self, data, conf):
         pass
+
+    def get_dependency_level(self, alpha: float) -> int:
```
[the remark this answers is not in the document]
Comment (jonyrock) [reply analytics/analytics/detectors/anomaly_detector.py]: ... by python doc string

File: analytics/analytics/detectors/anomaly_detector.py (hunk 1 above)
Comment (amper43) [reply analytics/analytics/detectors/anomaly_detector.py]: Specific type will be added in next iterations when format will be defined

**Pull request merged**